### PR TITLE
Stronger Cleanup Shard Snapshot Directory on Delete

### PR DESCRIPTION
**Pull request by original-brownbear**
* Use `RepositoryData` to clean up unreferenced `snap-${uuid}.dat` blobs
from shard directories (and index-N) and as a result also clean up data
blobs that are only referenced by them
* Stop cleaning up anything but index-N on shard snapshot creation to
align behavior of shard index-N handling with root path index-N
handling
  * This PR should by that also bring full alignment in index-N handling at the root and at the shard level with the only exception being the index-N rebuild code for when it's missing (I think we can actually remove that code, I don't see what purpose it really serves but we can look into that in a follow-up :)).

follow up to #44165

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -951,10 +953,26 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
...
+                blobsToDelete =
+                    blobs.keySet().stream().filter(blob -> blob.startsWith(SNAPSHOT_INDEX_PREFIX)).collect(Collectors.toList());
+            } catch (IOException e) {
+                throw new IndexShardSnapshotFailedException(shardId,
+                    "Failed to finalize snapshot creation [" + snapshotId + "] with shard index ["
+                        + indexShardSnapshotsFormat.blobName(indexGeneration) + "]", e);
+            }
             snapshotStatus.moveToDone(threadPool.absoluteTimeInMillis());
+            try {
+                shardContainer.deleteBlobsIgnoringIfNotExists(blobsToDelete);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I would prefer for the clean-up to happen before we move the snapshotStatus to done, as it used to be before. This ensures that there's no more activity on the repository from this data node once we've moved to DONE.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -951,10 +953,26 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
...
             }
-            // finalize the snapshot and rewrite the snapshot index with the next sequential snapshot index
-            finalizeShard(newSnapshotsList, fileListGeneration, blobs, "snapshot creation [" + snapshotId + "]", shardContainer,
-                shardId, snapshotId);
+            final String indexGeneration = Long.toString(fileListGeneration + 1);
+            final List<String> blobsToDelete;
+            try {
+                final BlobStoreIndexShardSnapshots updatedSnapshots = new BlobStoreIndexShardSnapshots(newSnapshotsList);
+                indexShardSnapshotsFormat.writeAtomic(updatedSnapshots, shardContainer, indexGeneration);
+                // Delete all previous index-N blobs
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: should we assert that it's indeed `previous` index-N blobs, i.e. that the numeric suffix is smaller than `indexGeneration`?

**Review by ywelsch — APPROVED** [review]
Thanks, I like that this moves all clean-up to the master.

I would like for @tlrx to have a look here too.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Right that was an oversight, moved it :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Sure why not, added it.

**Review by tlrx — APPROVED** [review]
LGTM, left a minor comment

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 3]
```diff
@@ -524,7 +526,7 @@ protected void doRun() {
                     if (indexMetaData != null) {
                         for (int shardId = 0; shardId < indexMetaData.getNumberOfShards(); shardId++) {
                             try {
-                                deleteShardSnapshot(indexId, new ShardId(indexMetaData.getIndex(), shardId), snapshotId);
+                                deleteShardSnapshot(repositoryData, indexId, new ShardId(indexMetaData.getIndex(), shardId), snapshotId);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: @tlrx I'm not sure after reading the code it seems like there's a bunch of exceptions that aren't caught here (all the `IndexShardSnapshotFailedException` in particular). As a matter of fact I'm not even sure we ever get a `SnapshotException` here (seems we're catching those down-stream where they could happen). Maybe clean this up in a follow-up (we def. should clean it up)?

[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: >  Maybe clean this up in a follow-up (we def. should clean it up)?

I'm good as long as it is cleaned up :)

**Comment by original-brownbear** [comment]
Thanks @tlrx + @ywelsch

**Pull request merged**